File: 215-KthLargestElementInAnArray.cpp
/*

Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.

For example,
Given [3,2,1,5,6,4] and k = 2, return 5.

Note: 
You may assume k is always valid, 1 ≤ k ≤ array's length.

*/

// Quick Select algo

class Solution {
public:
    int findKthLargest(vector<int>& nums, int n) {
        /*
        kth largest where k is [1,n], n - size of array
        kth largest = n-k+1 smallest number if indices from [1,n]
                    = n-k smallest number if indices from [0,n-1]
        */
        int k = nums.size() - n;
        random_shuffle(nums.begin(), nums.end()); // prevent worst case scenario of quick select
        int l = 0;
        int r = nums.size() - 1;
        while(true) {
            int piv = partition(nums, l, r);
            if(piv == k) break;
            else if (piv < k) l = piv+1;
            else r = piv - 1;
        }
        return nums[k];
    }
    
    int partition(vector<int>& nums, int l , int r) {
        int i = l;
        for (int j = i + 1; j <= r; ++j) {
            if (nums[j] <= nums[l]) {
                swap(nums[++i], nums[j]);
            }
        }
        swap(nums[i], nums[l]);
        return i;
    }
};

// using max heap
class Solution2 {
public:   
    inline int left(int idx) {
        return (idx << 1) + 1;
    }
    inline int right(int idx) {
        return (idx << 1) + 2;
    }
    void max_heapify(vector<int>& nums, int idx) {
        int largest = idx;
        int l = left(idx), r = right(idx);
        if (l < heap_size && nums[l] > nums[largest]) largest = l;
        if (r < heap_size && nums[r] > nums[largest]) largest = r;
        if (largest != idx) {
            swap(nums[idx], nums[largest]);
            max_heapify(nums, largest);
        }
    }
    void build_max_heap(vector<int>& nums) {
        heap_size = nums.size();
        for (int i = (heap_size >> 1) - 1; i >= 0; i--)
            max_heapify(nums, i);
    }
    int findKthLargest(vector<int>& nums, int k) {
        build_max_heap(nums);
        for (int i = 0; i < k; i++) {
            swap(nums[0], nums[heap_size - 1]);
            heap_size--;
            max_heapify(nums, 0);
        }
        return nums[heap_size];
    }
private:
    int heap_size;
};

// using stl priority queue (default max priority queue)
class Solution3 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int> pq(nums.begin(), nums.end());
        for (int i = 0; i < k - 1; i++)
            pq.pop(); 
        return pq.top();
    }
}; 

// using STL multi set
class Solution4 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        multiset<int> mset;
        int n = nums.size();
        for (int i = 0; i < n; i++) { 
            mset.insert(nums[i]);
            if (mset.size() > k)
                mset.erase(mset.begin());
        }
        return *mset.begin();
    }
};

File: java/215-KthLargestElementInAnArray.java
/*

Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.

For example,
Given [3,2,1,5,6,4] and k = 2, return 5.

Note: 
You may assume k is always valid, 1 ≤ k ≤ array's length.

*/

class Solution {
    public int findKthLargest(int[] nums, int k) {
        k = nums.length - k; // n-k+1 smallest element
        int l = 0, r = nums.length-1;
        while(true) {
            int piv  = partition(nums, l, r);
            if(piv == k) break;
            if(piv < k) l = piv + 1;
            else r = piv - 1;
        }
        return nums[k];
    }
    private int partition(int[] nums, int l, int r){
        int key = nums[l];
        int i = l;
        for(int j = i+1; j<=r; ++j) {
            if(nums[j] <= key) {
                swap(nums, ++i, j);
            }
        }
        swap(nums, i, l);
        return i;
    }
    private void swap(int[] nums, int i, int j){
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
        
    }
}

class Solution1 {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> maxpq = new PriorityQueue<Integer>(nums.length, new Comparator<Integer>(){
            public int compare(Integer a, Integer b) {
                if(a.compareTo(b) == 1) return -1;
                else if (a.compareTo(b) == -1) return 1;
                return 0;
            }
        });
        
        for(int n:nums) maxpq.offer(n);
        for(int i = 0; i<k-1; i++) maxpq.poll();
        return maxpq.peek();
    }
}
